src/Evolution/Systems/Ccz4/Ricci.cpp

```diff
@@ -0,0 +1,141 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/Systems/Ccz4/Ricci.hpp"
+
+#include <cstddef>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/ContainerHelpers.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace Ccz4 {
+template <size_t Dim, typename Frame, typename DataType>
+void ricci(
+    const gsl::not_null<tnsr::ii<DataType, Dim, Frame>*> result,
+    const tnsr::Ijj<DataType, Dim, Frame>& christoffel_second_kind,
+    const tnsr::iJkk<DataType, Dim, Frame>& d_conformal_christoffel_second_kind,
+    const tnsr::ii<DataType, Dim, Frame>& conformal_spatial_metric,
+    const tnsr::II<DataType, Dim, Frame>& inverse_conformal_spatial_metric,
+    const tnsr::ijj<DataType, Dim, Frame>& field_d,
+    const tnsr::iJJ<DataType, Dim, Frame>& field_d_up,
+    const tnsr::i<DataType, Dim, Frame>& field_p,
+    const tnsr::ij<DataType, Dim, Frame>& d_field_p) noexcept {
+  destructive_resize_components(result,
+                                get_size(get<0, 0>(conformal_spatial_metric)));
+  for (auto& component : (*result)) {
+    component = 0.0;
+  }
+
```

Comment:
Did you not want to use the tensor expressions for this computation? Up to you :)

Reply:
The implementation now uses `TensorExpression`s since https://github.com/sxs-collaboration/spectre/pull/3645 brought down the compilation time